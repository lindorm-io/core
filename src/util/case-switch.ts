import { camelCase, snakeCase } from "lodash";
import { isArrayStrict, isObjectStrict } from "./strict-type";

type Callback = (arg: string) => string;

const convertArrayValuesTo = (input: Array<string>, callback: Callback): Array<string> => {
  if (!isArrayStrict(input)) {
    throw new Error(`Invalid input [ ${typeof input} ]`);
  }

  const result: Array<string> = [];

  for (const value of input) {
    result.push(callback(value));
  }

  return result;
};

const convertObjectKeysTo = <Input extends Record<string, any>, Output extends Record<string, any>>(
  input: Input,
  callback: Callback,
): Output => {
  if (!isObjectStrict(input)) {
    throw new Error(`Invalid input [ ${typeof input} ]`);
  }

  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(input)) {
    if (isObjectStrict(value)) {
      result[callback(key)] = convertObjectKeysTo(value, callback);
    } else {
      result[callback(key)] = value;
    }
  }

  return result as Output;
};

export { camelCase };

export const camelArray = (input: Array<string>): Array<string> => {
  return convertArrayValuesTo(input, camelCase);
};

export const camelKeys = <Input extends Record<string, any>, Output extends Record<string, any>>(
  input: Input,
): Output => {
  return convertObjectKeysTo<Input, Output>(input, camelCase);
};

export const pascalCase = (string: string): string => {
  const [first, ...rest] = camelCase(string);
  return first.toUpperCase().concat(...rest);
};

export const pascalArray = (input: Array<string>): Array<string> => {
  return convertArrayValuesTo(input, pascalCase);
};

export const pascalKeys = <Input extends Record<string, any>, Output extends Record<string, any>>(
  input: Input,
): Output => {
  return convertObjectKeysTo<Input, Output>(input, pascalCase);
};

export { snakeCase };

export const snakeArray = (input: Array<string>): Array<string> => {
  return convertArrayValuesTo(input, snakeCase);
};

export const snakeKeys = <Input extends Record<string, any>, Output extends Record<string, any>>(
  input: Input,
): Output => {
  return convertObjectKeysTo<Input, Output>(input, snakeCase);
};
