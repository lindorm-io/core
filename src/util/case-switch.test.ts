import { camelArray, camelKeys, pascalArray, pascalCase, pascalKeys, snakeArray, snakeKeys } from "./case-switch";

describe("case-switch.ts", () => {
  describe("pascalCase", () => {
    test("should convert string to PascalCase", () => {
      expect(pascalCase("mock_string_one")).toBe("MockStringOne");
      expect(pascalCase("mockStringTwo")).toBe("MockStringTwo");
    });
  });

  describe("camelArray", () => {
    test("should only convert objects", () => {
      // @ts-ignore
      expect(camelArray("string")).toBe("string");
    });

    test("should convert object keys to camelCase", () => {
      expect(camelArray(["camelCase", "snake_case", "PascalCase"])).toMatchSnapshot();
    });
  });

  describe("camelKeys", () => {
    test("should only convert objects", () => {
      const string = "string";
      // @ts-ignore
      expect(camelKeys(string)).toBe(string);

      const error = new Error("error");
      expect(camelKeys(error)).toStrictEqual(error);

      const date = new Date();
      expect(camelKeys(date)).toStrictEqual(date);

      const array: Array<string> = ["array"];
      // @ts-ignore
      expect(camelKeys(array)).toStrictEqual(array);
    });

    test("should convert object keys to camelCase", () => {
      expect(
        camelKeys({
          snake_one: "one",
          snake_two: 2,
          snake_three: true,
          camelOne: ["mock"],
        }),
      ).toMatchSnapshot();
    });

    test("should convert objects within objects to camelCase", () => {
      expect(
        camelKeys({
          snake_one: "one",
          snake_two: {
            object_one: 1,
            object_two: "two",
            nested_object: {
              nested_one: "one",
              nested_two: ["array"],
            },
          },
        }),
      ).toMatchSnapshot();
    });
  });

  describe("snakeArray", () => {
    test("should convert object keys to snake_case", () => {
      expect(snakeArray(["camelCase", "snake_case", "PascalCase"])).toMatchSnapshot();
    });
  });

  describe("snakeKeys", () => {
    test("should convert object keys to snake_case", () => {
      expect(
        snakeKeys({
          camelOne: "one",
          camelTwo: 2,
          snake_one: "mock",
        }),
      ).toMatchSnapshot();
    });
  });

  describe("pascalArray", () => {
    test("should convert object keys to PascalCase", () => {
      expect(pascalArray(["camelCase", "snake_case", "PascalCase"])).toMatchSnapshot();
    });
  });

  describe("pascalKeys", () => {
    test("should convert object keys to PascalCase", () => {
      expect(
        pascalKeys({
          camelOne: "one",
          snake_one: "one",
        }),
      ).toMatchSnapshot();
    });
  });
});
