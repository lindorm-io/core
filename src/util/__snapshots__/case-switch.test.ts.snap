// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`case-switch.ts camelArray should convert object keys to camelCase 1`] = `
Array [
  "camelCase",
  "snakeCase",
  "pascalCase",
]
`;

exports[`case-switch.ts camelKeys should convert object keys to camelCase 1`] = `
Object {
  "camelOne": Array [
    "mock",
  ],
  "snakeOne": "one",
  "snakeThree": true,
  "snakeTwo": 2,
}
`;

exports[`case-switch.ts camelKeys should convert objects within objects to camelCase 1`] = `
Object {
  "snakeOne": "one",
  "snakeTwo": Object {
    "nestedObject": Object {
      "nestedOne": "one",
      "nestedTwo": Array [
        "array",
      ],
    },
    "objectOne": 1,
    "objectTwo": "two",
  },
}
`;

exports[`case-switch.ts pascalArray should convert object keys to PascalCase 1`] = `
Array [
  "CamelCase",
  "SnakeCase",
  "PascalCase",
]
`;

exports[`case-switch.ts pascalKeys should convert object keys to PascalCase 1`] = `
Object {
  "CamelOne": "one",
  "SnakeOne": "one",
}
`;

exports[`case-switch.ts snakeArray should convert object keys to snake_case 1`] = `
Array [
  "camel_case",
  "snake_case",
  "pascal_case",
]
`;

exports[`case-switch.ts snakeKeys should convert object keys to snake_case 1`] = `
Object {
  "camel_one": "one",
  "camel_two": 2,
  "snake_one": "mock",
}
`;
